src/libraries/System.Private.CoreLib/src/System/Collections/Generic/ArraySortHelper.cs

```diff
@@ -406,147 +379,158 @@ private static int BinarySearch(T[] array, int index, int length, T value)
+
                 if (partitionSize <= IntrospectiveSortUtilities.IntrosortSizeThreshold)
                 {
-                    if (partitionSize == 1)
-                    {
-                        return;
-                    }
                     if (partitionSize == 2)
                     {
-                        SwapIfGreater(keys, lo, hi);
+                        SwapIfGreaterKnownDistinct(ref keys[0], ref keys[1]);
                         return;
                     }
+
                     if (partitionSize == 3)
                     {
-                        SwapIfGreater(keys, lo, hi - 1);
-                        SwapIfGreater(keys, lo, hi);
-                        SwapIfGreater(keys, hi - 1, hi);
+                        ref T hiRef = ref keys[2];
+                        ref T him1Ref = ref keys[1];
+                        ref T loRef = ref keys[0];
+
+                        SwapIfGreaterKnownDistinct(ref loRef, ref him1Ref);
+                        SwapIfGreaterKnownDistinct(ref loRef, ref hiRef);
+                        SwapIfGreaterKnownDistinct(ref him1Ref, ref hiRef);
                         return;
                     }
 
-                    InsertionSort(keys[lo..(hi+1)]);
+                    InsertionSort(keys[0..partitionSize]);
                     return;
                 }
 
                 if (depthLimit == 0)
                 {
-                    HeapSort(keys[lo..(hi+1)]);
+                    HeapSort(keys[0..partitionSize]);
                     return;
                 }
                 depthLimit--;
 
-                int p = PickPivotAndPartition(keys[lo..(hi+1)]);
+                int p = PickPivotAndPartition(keys[0..partitionSize]);
 
                 // Note we've already partitioned around the pivot and do not have to move the pivot again.
-                IntroSort(keys[(p+1)..(hi+1)], depthLimit);
+                IntroSort(keys[(p+1)..partitionSize], depthLimit);
                 hi = p - 1;
             }
         }
 
         private static int PickPivotAndPartition(Span<T> keys)
         {
-            Debug.Assert(!keys.IsEmpty);
-
-            int lo = 0;
-            int hi = keys.Length - 1;
-
-            // Compute median-of-three.  But also partition them, since we've done the comparison.
-            int middle = lo + ((hi - lo) / 2);
-
-            // Sort lo, mid and hi appropriately, then pick mid as the pivot.
-            SwapIfGreater(keys, lo, middle);  // swap the low with the mid point
-            SwapIfGreater(keys, lo, hi);   // swap the low with the high
-            SwapIfGreater(keys, middle, hi); // swap the middle with the high
-
-            T pivot = keys[middle];
-            Swap(keys, middle, hi - 1);
-            int left = lo, right = hi - 1;  // We already partitioned lo and hi and put the pivot in hi - 1.  And we pre-increment & decrement below.
-
-            while (left < right)
+            Debug.Assert(keys.Length >= 2);
```

Comment:
Do we get here only for `keys.Length >= IntrosortSizeThreshold` ?

I believe it means that all `SwapIfGreater` calls can be `SwapIfGreaterKnownDistinct`, all `Swap` calls can be `SwapKnownDistinct`, and all both the `SwapIfGreater` and `SwapIfGreater` methods can be deleted.

The same optimization should be possible in the other clones.

Reply:
Hmm, I think you're right.  I'll fix that up.